lib/team.rb

```diff
@@ -19,4 +19,48 @@ def team_info
       "link" => link }
   end
 
+  def best_season
+    seasons_by_win_percentage.max_by do |season, win_percentage|
+      win_percentage
+    end.first
+  end
+
+  def worst_season
+    seasons_by_win_percentage.min_by do |season, win_percentage|
+      win_percentage
+    end.first
+  end
+
+  def seasons_by_win_percentage
+    seasons_by_win_percentage = {}
+    games_by_season.each do |season, games|
+      seasons_by_win_percentage[season] = win_percentage(games)
+    end
+    seasons_by_win_percentage
+  end
+
+  def games_by_season
+    team_games.group_by do |team_game|
+      games.find do |game|
+        team_game.game_id == game.game_id
+      end.season
+    end
+  end
+
+  #Doesn't feels like it belong in this class, but maybe a module
+  def win_percentage(games)
+    wins = games.count do |game|
+      game.result == "WIN"
+    end
+    (wins / games.count.to_f).round(2)
+  end
+
+  def team_games
+    @manager.tracker.game_team_manager.find_games_by_team(team_id)
+  end
+
+  def games
+    @manager.tracker.game_manager.find_games_by_team(team_id)
+  end
```

Comment:
I think a team should know what games it plays in, but this feels like it reaches through a couple classes. Does it abide by the Law of Demeter?

Reply:
It seems appropriate to me that the win_percentage should be calculated at the individual Team level because every team will have a win percentage. If this logic is needed outside of this class then I would agree it would fit in a module.

find_games_by_team(team_id) could potentially be extracted and moved to the Groupable module removing the need to reach deeply, but we would need to provide a means of the module accessing the "games" instance variable. Maybe something like:

def find_games_by_team(data, team_id)
  self.send(data).find_all do |game|
    game.away_team_id == team_id || game.home_team_id == team_id
  end
end

Not sure if that works though...

Reply:
Where "data" is the symbol :games